container-search/src/main/java/com/yahoo/search/dispatch/SearchPath.java

```diff
@@ -0,0 +1,253 @@
+// Copyright 2018 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+package com.yahoo.search.dispatch;
+
+import com.google.common.collect.ImmutableMap;
+import com.yahoo.collections.Pair;
+import com.yahoo.search.dispatch.SearchCluster.Group;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+/**
+ * Utility class for parsing model.searchPath and filtering a search cluster
+ * based on it.
+ *
+ * @author ollivir
+ */
+public class SearchPath {
+    /**
+     * Parse the search path and select nodes from the given cluster based on it.
+     *
+     * @param searchPath
+     *            unparsed search path expression (see: model.searchPath in Search
+     *            API reference)
+     * @param cluster
+     *            the search cluster from which nodes are selected
+     * @throws InvalidSearchPathException
+     *             if the searchPath is malformed
+     * @return list of nodes chosen with the search path, or an empty list in which
+     *         case some other node selection logic should be used
+     */
+    public static List<SearchCluster.Node> selectNodes(String searchPath, SearchCluster cluster) {
+        Optional<SearchPath> sp = SearchPath.fromString(searchPath);
+        if (sp.isPresent()) {
+            return sp.get().mapToNodes(cluster);
+        } else {
+            return Collections.emptyList();
+        }
+    }
+
+    public static Optional<SearchPath> fromString(String path) {
+        if (path == null || path.isEmpty()) {
+            return Optional.empty();
+        }
+        if (path.indexOf(';') >= 0) {
+            return Optional.empty(); // multi-level not supported at this time
+        }
+        try {
+            SearchPath sp = parseElement(path);
+            if (sp.isEmpty()) {
+                return Optional.empty();
+            } else {
+                return Optional.of(sp);
+            }
+        } catch (NumberFormatException | InvalidSearchPathException e) {
+            throw new InvalidSearchPathException("Invalid search path: " + path);
+        }
+    }
+
+    private final List<Part> parts;
+    private final Integer row;
+
+    private SearchPath(List<Part> parts, Integer row) {
+        this.parts = parts;
```

Comment:
I have on my todo list to rename row->group, part->node ("searchPath" is older than those). Perhaps you can do that here.

Reply:
Gladly. It wasn't at all obvious which way the terminology maps.